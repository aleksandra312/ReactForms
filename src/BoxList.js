import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from "uuid";

const BoxList = () => {
  const [boxes, setBoxes] = useState([]);
  const addBox = (newBox) => {
    setBoxes((boxes) => [...boxes, { ...newBox, id: uuid() }]);
  };
  const removeBox = (id) => {
    setBoxes((boxes) => boxes.filter((b) => b.id !== id));
  };

  return (
    <div>
      <h3>Box List</h3>
      <NewBoxForm addBox={addBox} />
      <div>
        {boxes.map(({ id, color, width, height }) => (
          <Box
            id={id}
            color={color}
            width={width}
            height={height}
            key={id}
            removeBox={removeBox}
          />
        ))}
      </div>
    </div>
  );
};

export default BoxList;
