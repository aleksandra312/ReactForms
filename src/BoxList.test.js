import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, height = "5", width = "5", color = "yellow") {
  const colorInput = boxList.getByLabelText("Color");
  const widthInput = boxList.getByLabelText("Width");
  const heightInput = boxList.getByLabelText("Height");

  fireEvent.change(colorInput, { target: { value: color } });
  fireEvent.change(widthInput, { target: { value: width } });
  fireEvent.change(heightInput, { target: { value: height } });

  const addBtn = boxList.getByText("Add Box");
  fireEvent.click(addBtn);
}

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function () {
  const boxList = render(<BoxList />);
  expect(boxList.queryByText("X")).not.toBeInTheDocument();

  addBox(boxList);
  expect(boxList.getByText("X")).toBeInTheDocument();
});

it("can remove a box", function () {
  const boxList = render(<BoxList />);
  addBox(boxList);

  const removeBtn = boxList.getByText("X");
  fireEvent.click(removeBtn);
  expect(removeBtn).not.toBeInTheDocument();
});
