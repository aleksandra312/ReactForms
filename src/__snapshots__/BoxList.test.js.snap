// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Box List
    </h3>
    <form>
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        placeholder="box color"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        max="30"
        min="5"
        name="width"
        type="range"
        value=""
      />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        max="30"
        min="5"
        name="height"
        type="range"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
